### ISSUE TITLE
confused by the update policy. 

### ISSUE THREAD

**Issue by fuyunfei** [megvii-research/ICCV2019-LearningToPaint#49]
in the update_policy() :
`cur_q, step_reward = self.evaluate(state, action) target_q += step_reward.detach() value_loss = criterion(cur_q, target_q)`

it's quite confusing .. so the value_loss = discount*(self.critic(St+1)+reward(St+1)) -self.critic(St) ??

shouldn't it be :  Value_loss =  discount*(self.critic(St+1)) + reward(St) - self.critic(St) ?

**Comment by fuyunfei** [comment]
`        cur_q, step_reward = self.evaluate(state, action)
        target_q += step_reward.detach()
        
        value_loss = criterion(cur_q, target_q)`

it's quite confusing ..   so the  value_loss  = discount*(self.critic(St+1)+reward(St+1))   -self.critic(S)   ??

**Comment by fuyunfei** [comment]
非常谢谢你的回答，可是这里的target_q  =  (Q + gan_reward) ,  即  

133行：  target_q =  (self.critic(St+1) + reward(St+1) ) * discount, 
135行： cur_q = self.critic(St) + reward(St)
136行：  target_q=  (self.critic(St+1) + reward(St+1) ) * discount  + reward(St),

所以：  V_loss =  target_q - cur_q =    (self.critic(St+1) + reward(St+1) ) * discount  - self.critic(St) .  

好像不一致唉，是我哪里理解错了吗 ？  
谢谢~

**Comment by hzwer** [comment]
这个公式是对的，V_loss ～ target_q - cur_q ～ (self.critic(St+1) + reward(St+1) ) * discount - self.critic(St) .
可能是因为我 evaluate 算 Q 函数的时候，把单前步的 reward 分出来了，所以形式上有点奇怪

**Comment by fuyunfei** [comment]
对吗 ... 
因为
![image](https://user-images.githubusercontent.com/9105298/111735497-148b5380-88b7-11eb-85c4-6b02e387925d.png)

所以，
 V_loss ～ target_q - cur_q ～ self.critic(St+1)  * discount + reward(St)  - self.critic(St)  而不是 (self.critic(St+1) + reward(St+1) ) * discount - self.critic(St) . 

如果严格按强化学习的公式，gan_reward + critic 应该整体看作一个 Q:

那是不是应该是
Q(S_t,a_t) = reward(St,at) + discount * V(St+1)   

而不是目前code里面的：
Q(S_t,a_t) = reward(St,at) + V(St)    ？ 


Correct me if I'm wrong .... ..   
Thanks .

**Comment by hzwer** [comment]
self.critic(St+1) * discount + reward(St) - self.critic(St)
和 (self.critic(St+1) + reward(St+1) ) * discount - self.critic(St)
在本质上没有任何区别呀 :)
discount 加不加在第一个 reward 上，相当于折扣回报系数标号变一下而已

**Comment by fuyunfei** [comment]
got it , 非常感谢！
